cmd/restic/cmd_rewrite.go

```diff
@@ -0,0 +1,296 @@
+				continue
+			}
+			newNode := node
+			newID, err := filterNode(ctx, repo, path, *node.Subtree, checkExclude, saveTreeFunc)
+			if err != nil {
+				return nodeID, err
+			}
+			if newID == *node.Subtree {
+				newTree.Insert(node)
+			} else {
+				changed = true
+				newNode.Subtree = new(restic.ID)
+				*newNode.Subtree = newID
+				newTree.Insert(newNode)
+			}
+		} else {
+			Verbosef("excluding %s\n", path)
+			changed = true
+		}
+	}
+
+	if changed {
+		// Save new tree
+		newTreeID, err := saveTreeFunc(newTree)
+		debug.Log("filterNode: save new tree for %s as %v\n", nodepath, newTreeID)
+		return newTreeID, err
+	}
+
+	return nodeID, nil
+}
+
+func collectRejectFuncsForRewrite(opts RewriteOptions) (fs []RejectByNameFunc, err error) {
+	//TODO: merge with cmd_backup
+
+	// add patterns from file
+	if len(opts.ExcludeFiles) > 0 {
+		excludes, err := readExcludePatternsFromFiles(opts.ExcludeFiles)
+		if err != nil {
+			return nil, err
+		}
+		opts.Excludes = append(opts.Excludes, excludes...)
+	}
+
+	if len(opts.InsensitiveExcludes) > 0 {
+		fs = append(fs, rejectByInsensitivePattern(opts.InsensitiveExcludes))
+	}
+
+	if len(opts.Excludes) > 0 {
+		fs = append(fs, rejectByPattern(opts.Excludes))
+	}
+
+	return fs, nil
+}
+
+func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *restic.Snapshot, opts RewriteOptions, gopts GlobalOptions) (bool, error) {
+	if sn.Tree == nil {
+		return false, errors.Errorf("snapshot %v has nil tree", sn.ID().Str())
+	}
+
+	var saveTreeFunc saveTreeFunction
+	if !opts.DryRun {
+		saveTreeFunc = func(tree *restic.Tree) (restic.ID, error) {
+			return repo.SaveTree(ctx, tree)
+		}
+	} else {
+		saveTreeFunc = func(tree *restic.Tree) (restic.ID, error) {
+			return restic.ID{0}, nil
+		}
+	}
+
+	rejectByNameFuncs, err := collectRejectFuncsForRewrite(opts)
+	if err != nil {
+		return false, err
+	}
+
+	checkExclude := func(nodepath string) bool {
+		for _, reject := range rejectByNameFuncs {
+			if reject(nodepath) {
+				return true
+			}
+		}
+		return false
+	}
+
+	filteredTree, err := filterNode(ctx, repo, "/", *sn.Tree, checkExclude, saveTreeFunc)
+
+	if err != nil {
+		return false, err
+	}
+
+	if filteredTree == *sn.Tree {
+		debug.Log("Snapshot not touched\n")
+		return false, nil
+	}
+
+	debug.Log("Snapshot modified\n")
+	if opts.DryRun {
+		Printf("Will modify snapshot: %s\n", sn.String())
+		return true, nil
+	}
+
+	err = repo.Flush(ctx)
+	if err != nil {
+		return false, err
+	}
+	err = repo.SaveIndex(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// Retain the original snapshot id over all tag changes.
+	if sn.Original == nil {
+		sn.Original = sn.ID()
+	}
+	*sn.Tree = filteredTree
+
+	if !opts.Inplace {
+		sn.AddTags([]string{"rewrite"})
+	}
+
+	// Save the new snapshot.
+	id, err := repo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+	if err != nil {
+		return false, err
+	}
+	err = repo.Flush(ctx)
+	if err != nil {
+		return true, err
+	}
+
+	if opts.Inplace {
```

Comment:
I would suggest to remove the deletion code totally from `cmd_rewrite.go` and instead call `runForget` from `forget.go` with the snapshot-IDs to remove.

That way you will not have any critical code that removes anything here, but keep that "burden" in exclusive pieces of code (which is well-tested and can be optimized or maybe turned into 2-phase-compaction later more easily).

Also allows to add a `--prune` flag here and pass this directly to `runForget`; we should then also add the `addPruneFlags()` call once #2718 is merged.

Reply:
Chaining the commands is probably more complex than directly removing the snapshot. In particular the error handling, log output etc. would have to be coordinated between the commands.